Add modular obsoletes support to libdnf.

kontura opened the pull request:
Couple of things I am not certain about:
- when eol switches not only stream but also a module the original module is disabled.
- when eols create a chain (nodejs:5 is obosoleted by nodejs:10 and nodejs:10 is obsoleted by nodejs:11) shloud we (can we?) resolve it in one transaction? Currently only one eol is applied per transaction.
- I have also noticed we named the modular metadata eol (end of life) but libdnf option is `module_allow_obsoletes` would `module_allow_eols` be more consistent?

Requires: https://github.com/fedora-modularity/libmodulemd/pull/476
Is requried by: https://github.com/rpm-software-management/dnf/pull/1641
Tests: https://github.com/rpm-software-management/ci-dnf-stack/pull/869

kontura committed:
Add new option module_obsoletes

= changelog =
msg: Add a new option module_obsoletes
type: enhancement

kontura committed:
Add getter for modular obsoletes from ModuleMetadata

kontura committed:
Add counting of modular stream changes

Also extend enable, disable and reset methods to allow not to
count their specific change, this is needed for obsoletes which
should not count into the limit of one change per stream.

j-mracek commented on libdnf/module/ModulePackageContainer.cpp:
```diff
@@ -1691,4 +1704,55 @@ void ModulePackageContainer::updateFailSafeData()
+            auto moduleStream = modulemd_eol_get_obsoleted_by_module_stream(e);
+
+            if (moduleName && moduleStream) {
+                enable(moduleName, moduleStream, false);
```
What about when module `moduleName` is disabled? Then I would suggest that obsolete should not be applied.

dmach replied:
That sounds like a "conflicting request" in transaction.
DNF should error out on that.

kontura replied:
But then every transaction would throw the error until the user either enables the target module, disables the obsoleted stream or the metadata are changed.

Conan-Kudo commented:
> when eol switches not only stream but also a module the original module is disabled.

This _sounds_ like what should happen, but could you clarify what you mean here?

> when eols create a chain (nodejs:5 is obosoleted by nodejs:10 and nodejs:10 is obsoleted by nodejs:11) shloud we (can we?) resolve it in one transaction? Currently only one eol is applied per transaction.

We should resolve *all* of them in one transaction.

> I have also noticed we named the modular metadata eol (end of life) but libdnf option is module_allow_obsoletes would module_allow_eols be more consistent?

It should just be `module_obsoletes` similar to the regular `obsoletes` one.

dmach commented on libdnf/conf/ConfigMain.cpp:
```diff
@@ -430,6 +432,8 @@ ConfigMain::Impl::Impl(Config & owner)
     owner.optBinds().add("comment", comment);
     owner.optBinds().add("ignorearch", ignorearch);
     owner.optBinds().add("module_platform_id", module_platform_id);
+    owner.optBinds().add("module_allow_obsoletes", module_allow_obsoletes);
```
How about `module_allow_eol`?
Eol is superset of obsoletes functionality and refers to the modulemd-eol documents.
On the other hand module_allow_obsoletes is easier to read :(

j-mracek replied:
The proposed name `module_obsoletes` from @Conan-Kudo is a better choice from my point of view.

j-mracek commented:
> > when eols create a chain (nodejs:5 is obosoleted by nodejs:10 and nodejs:10 is obsoleted by nodejs:11) shloud we (can we?) resolve it in one transaction? Currently only one eol is applied per transaction.
> 
> We should resolve _all_ of them in one transaction.
> 
That good idea, but the problem is in time line. For Fedora 33 we will deliver module_obsoletes, for Fedora 34 a new modular solver with a new context modular. To resolve _all_ obsoletes will require change in modular solver, but it is planned for Fedora 34. Probably we need a time machine and multidimensional universe to achieve it.

kontura committed:
Add new API applyObsoletes() function to apply modular obsoletes

= changelog =
msg: Add new API applyObsoletes() function to apply modular obsoletes
type: enhancement

kontura commented:
Even thought it looks like we won't use this for F33 and we can integrate it more in F34 with a new modular solver I still wanted to push the update since I had it basically finished.

I have renamed the options to: `module_obsoletes` and `module_stream_switch` and I have simplified it, moving most of the logic to libmodulemd. Also the obsoletes are now applied based on an optional parameter to `filter_modules` which is more efficient because we don't have to compute the modules twice.

j-mracek commented:
I suggest that the present implementation will be compatible with a new solver (with support of new contexts). The solver will only allow additional possibility how to resolve obsoletes.

j-mracek commented on libdnf/module/modulemd/ModuleMetadata.cpp:
```diff
@@ -193,4 +193,32 @@ void ModuleMetadata::reportFailures(const GPtrArray *failures)
 
+ModulemdObsoletes * ModuleMetadata::getNewestActiveObsolete(ModulePackage *p)
+{
+    ModulemdModule * myModule = modulemd_module_index_get_module(resultingModuleIndex, p->getName().c_str());
```
Minor comments:
I am not happy when single letter variable is used. See libsolv and p and pp variables in macros and both have type int.
Also call p->getName().c_str() is not much efficient. Internally  it calls `modulemd_module_stream_get_module_name(mdStream)`. When it returns null it is converted into empty string. and empty string is not NULL.

j-mracek commented on libdnf/module/modulemd/ModuleMetadata.cpp:
```diff
@@ -193,4 +193,32 @@ void ModuleMetadata::reportFailures(const GPtrArray *failures)
+                                                                          p->getArchCStr(),
+                                                                          &error);
+    if (error) {
+        auto logger(libdnf::Log::getLogger());
```
When the error appears? When I ask for data that are not present? Like I have module `nodejs`, stream `12` but no obsolete data available. Or when I ask for data for stream that doesn't exist at all? Like stream `sssss`.

kontura replied:
When you ask for a stream that doesn't exists. It shouldn't happen.

j-mracek commented on libdnf/module/modulemd/ModuleMetadata.cpp:
```diff
@@ -193,4 +193,32 @@ void ModuleMetadata::reportFailures(const GPtrArray *failures)
+    }
+
+    GError *error = NULL;
+    ModulemdModuleStream * myStream = modulemd_module_get_stream_by_NSVCA(myModule,
```
ModulePackage contains ModulemdModuleStream, why you search for it? Just to need to be a friend. Or did I overlooked something?

kontura replied:
Unfortunately the `ModulemdModuleStream` inside this `ModulePackage` was constructed with metadata from only one repository and the obsoletes may be missing. Its mentioned in a comment before this function is called in `applyObsoletes`. We have to find the stream in `pImpl->moduleMetadata` instance from `ModulePackageContainer` because it contains all available module metadata merged together.

j-mracek replied:
Ok, thanks for an explanation.

kontura commented:
I have rebased it and added a commit which enables obsoletes for microdnf, which now support modules. I am not sure if it could not somehow break packagekit.

TomasTomecek commented:
/packit build
no longer our infra problem :> 
```
No matching package to install: 'pkgconfig(modulemd-2.0) >= 2.10.0'
Not all dependencies satisfied
```

kontura commented:
Updated spec require for libmodulemd-2.10.0-2.

I have also removed the commit which enabled obsoletes in microdnf.

j-mracek commented on libdnf/module/ModulePackageContainer.hpp:
```diff
@@ -142,7 +142,7 @@ struct ModulePackageContainer
     * @return bool
     */
-    bool enable(const std::string &name, const std::string &stream);
+    bool enable(const std::string &name, const std::string &stream, const bool count = true);
```
@kontura Please could you add a description to the `count` argument not only in this method?

kontura replied:
Added a description, please check if its understandable.

j-mracek replied:
What about ?
 ```
When the count parameter is set to false the change will be not count towards the limit of module state modification
```
But I am not sure.

j-mracek replied:
The documentation should be also enhanced in disable and reset methods.

kontura replied:
ok, updated the documentation for enable, disable and reset with basically the version you suggested.

j-mracek commented on VERSION.cmake:
```diff
@@ -1,6 +1,6 @@
 set (DEFAULT_LIBDNF_MAJOR_VERSION 0)
 set (DEFAULT_LIBDNF_MINOR_VERSION 55)
-set (DEFAULT_LIBDNF_MICRO_VERSION 2)
+set (DEFAULT_LIBDNF_MICRO_VERSION 3)
```
[only a suggestion] May be I would recommend to change minor version rather than patch version. But this not an issue.

kontura committed:
Add an option to apply modular obsoletes to modular_filtering

This option is for now permanenly turned of for context part of libdnf
(microdnf and packagekit).

= changelog =
msg: Extend filter_modules with an optional parameter to enable applying modular obsoletes
type: enhancement

j-mracek commented:
@rh-atomic-bot r+

j-mracek commented:
Tested with `libmodulemd-2.12.0-1.fc32.x86_64`
